opsi/manager/pipeline.py

```diff
@@ -24,12 +27,138 @@ class Connection(NamedTuple):
 Links = Dict[str, Connection]
 
 
+class SingleRunPerformance(NamedTuple):
+    nodes: Dict[str, float]
+
+
+class Performance:
+    # All floats here are time durations measured in seconds
+    __slots__ = ("nodes", "node_types", "pipeline", "sum_nodes", "current_run")
+
+    def __init__(self, nodes: Dict[str, "Node"]):
+        self.nodes: Dict[str, List[float]] = {id: [] for id in nodes.keys()}
+        self.node_types: Dict[str, str] = {
+            id: node.func_type.type for id, node in nodes.items()
+        }
+        self.pipeline: List[float] = []
+        self.sum_nodes: List[float] = []
+        self.current_run: SingleRunPerformance = None
+
+    def new_run(self):
+        self.current_run = SingleRunPerformance(nodes=dict.fromkeys(self.nodes.keys()))
+
+    def log_node_run(self, id, time):
+        self.current_run.nodes[id] = time
+
+    def finalize_run(self, pipeline_perf):
+        # filter out None and zero
+        self.sum_nodes.append(math.fsum(filter(None, self.current_run.nodes.values())))
+
+        for id, data in self.nodes.items():
+            entry = self.current_run.nodes[id]
+            if entry is not None:
```

Comment:
Why not just `if not entry:` if you're filtering none and zero?

Reply:
I don't think we should filter zeros, if it's valid data.
There wouldn't be zeros normally, except maybe if the time is imprecise, but perf_counter should be precise.

Reply:
from comment: `# filter out None and zero`

Reply:
That comment is referring to the behavior of `filter(None, ...)`, which is, filtering out falsey values: https://docs.python.org/3/library/functions.html#filter
> If function is None, the identity function is assumed, that is, all elements of iterable that are false are removed.
> `(item for item in iterable if item)`
Zeros don't affect the sum of the pipeline run, but they do affect the node average.